language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -239,29 +246,41 @@ substitution, anything).
 computeCek
     :: (GShow uni, GEq uni, DefaultUni <: uni, Closed uni, uni `Everywhere` ExMemoryUsage)
     => Context uni -> WithMemory Term uni -> CekM uni (Plain Term uni)
-computeCek con t@(TyInst _ body ty) = do
+computeCek ctx t@(TyInst _ body ty) = do
     spendBudget BTyInst t (ExBudget 1 1) -- TODO
-    computeCek (FrameTyInstArg ty : con) body
-computeCek con t@(Apply _ fun arg) = do
+    computeCek (FrameTyInstArg ty : ctx) body
+computeCek ctx t@(Apply _ fun arg) = do
     spendBudget BApply t (ExBudget 1 1) -- TODO
     varEnv <- getVarEnv
-    computeCek (FrameApplyArg varEnv arg : con) fun
-computeCek con t@(IWrap ann pat arg term) = do
+    computeCek (FrameApplyArg varEnv arg : ctx) fun
+computeCek ctx t@(IWrap ann pat arg term) = do
     spendBudget BIWrap t (ExBudget 1 1) -- TODO
-    computeCek (FrameIWrap ann pat arg : con) term
-computeCek con t@(Unwrap _ term) = do
+    computeCek (FrameIWrap ann pat arg : ctx) term
+computeCek ctx t@(Unwrap _ term) = do
     spendBudget BUnwrap t (ExBudget 1 1) -- TODO
-    computeCek (FrameUnwrap : con) term
-computeCek con tyAbs@TyAbs{}       = returnCek con tyAbs
-computeCek con lamAbs@LamAbs{}     = returnCek con lamAbs
-computeCek con constant@Constant{} = withVarEnv mempty $ returnCek con constant
-computeCek con bi@Builtin{}        = returnCek con bi
+    computeCek (FrameUnwrap : ctx) term
+computeCek ctx tyAbs@TyAbs{}       = returnCek ctx tyAbs
+computeCek ctx lamAbs@LamAbs{}     = returnCek ctx lamAbs
+computeCek ctx constant@Constant{} = returnCek ctx constant
+computeCek ctx t@(ApplyBuiltin _ bn tys args) =
+    -- FIXME (FAO @reactormonk): I've just re-used BApply for costing
+    -- the use of FrameApplyBuiltin.  Maybe we should add a new
+    -- constructor to `ExBudgetCategory` for that.
+    case args of
+        [] -> do
+            spendBudget BApply t (ExBudget 1 1)
+            varEnv <- getVarEnv
+            applyBuiltin varEnv ctx bn tys []
```

Comment:
I think we don't really support this case? At least costing doesn't work correctly here IIRC. I think we could throw an error here and think about this case later.

Reply:
> A thought: we could use `NonEmpty` for the builtin arguments and give an error way back in the parser if there's no arguments at all.

OK,  I'll try that.  I did wonder if we could encode the type and term arities at the type level so that we could _never_ have an application that was incorrectly instantiated/applied, but I thought that was too ambitious at this point.  I'm also not sure how that would interact with CBOR decoding, for example.

Reply:
Yeah, matching arity sounds annoying, but non-emptiness is a pretty easy thing to add if you know it holds.

Decoding is a good example, though: if we made nullary builtins work, it would be nice if we didn't have to change the serialization format...

Reply:
I spent quite a long time on this and then gave up.  It requires changes to a lot of code, including the prettyprinter, the typechecker and so on. It was just making a horrible mess of everything, so I've made the parser and the evaluators (but not the CBOR decoder, yet) reject nullary builtin applications.  I might revisit this later, but not right now.

Reply:
I made an issue about nullary builtins at SCP-779.